66_HLSLBxDFTests/app_resources/tests.hlsl

```diff
@@ -0,0 +1,455 @@
+            sy = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy + float32_t2(0,base_t::rc.h));
+        }
+        if NBL_CONSTEXPR_FUNC (is_microfacet_brdf_v<BxDF>)
+        {
+            s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy, cache);
+            sx = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy + float32_t2(base_t::rc.h,0), dummy);
+            sy = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy + float32_t2(0,base_t::rc.h), dummy);
+        }
+        if NBL_CONSTEXPR_FUNC (is_basic_bsdf_v<BxDF>)
+        {
+            s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u);
+            sx = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u + float32_t3(base_t::rc.h,0,0));
+            sy = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u + float32_t3(0,base_t::rc.h,0));
+        }
+        if NBL_CONSTEXPR_FUNC (is_microfacet_bsdf_v<BxDF>)
+        {
+            s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u, cache);
+            float32_t3 ux = base_t::rc.u + float32_t3(base_t::rc.h,0,0);
+            sx = base_t::bxdf.generate(base_t::anisointer, ux, dummy);
+            float32_t3 uy = base_t::rc.u + float32_t3(0,base_t::rc.h,0);
+            sy = base_t::bxdf.generate(base_t::anisointer, uy, dummy);
+        }
+        
+        if NBL_CONSTEXPR_FUNC (is_basic_brdf_v<BxDF> || is_basic_bsdf_v<BxDF>)
+        {
+            pdf = base_t::bxdf.quotient_and_pdf(s, base_t::isointer);
+            brdf = float32_t3(base_t::bxdf.eval(s, base_t::isointer));
+        }
+        if NBL_CONSTEXPR_FUNC (is_microfacet_brdf_v<BxDF> || is_microfacet_bsdf_v<BxDF>)
+        {
+            if NBL_CONSTEXPR_FUNC (aniso)
+            {
+                pdf = base_t::bxdf.quotient_and_pdf(s, base_t::anisointer, cache);
+                brdf = float32_t3(base_t::bxdf.eval(s, base_t::anisointer, cache));
+            }
+            else
+            {
+                iso_cache isocache = (iso_cache)cache;
+                pdf = base_t::bxdf.quotient_and_pdf(s, base_t::isointer, isocache);
+                brdf = float32_t3(base_t::bxdf.eval(s, base_t::isointer, isocache));
+            }
+        }
+
+        // get jacobian
+        float32_t2x2 m = float32_t2x2(sx.TdotL - s.TdotL, sy.TdotL - s.TdotL, sx.BdotL - s.BdotL, sy.BdotL - s.BdotL);
+        float det = nbl::hlsl::determinant<float32_t2x2>(m);
+
+        return float32_t4(nbl::hlsl::abs<float32_t3>(pdf.value() - brdf), nbl::hlsl::abs<float>(det*pdf.pdf/s.NdotL) * 0.5);
+    }
+
+    static STestMeta run(uint32_t2 seed)
+    {
+        this_t t;
+        t.init(seed);
+        if NBL_CONSTEXPR_FUNC (is_microfacet_brdf_v<BxDF> || is_microfacet_bsdf_v<BxDF>)
+            t.template initBxDF<aniso>(t.rc);
+        else
+            t.initBxDF(t.rc);
+
+        t.meta.result = t.test();
+        t.meta.testName = "u offset";
+        return t.meta;
+    }
```

Comment:
setting those thousands of times will be expensive, function should be `void` and not return anything, callback instead of meta